File: packages/server/src/services/local-scan.service.ts
import { readdir, access } from 'fs/promises';
import { join, resolve } from 'path';
import { execSync } from 'child_process';
import { createLogger } from '../utils/logger.js';
import { getErrorMessage } from '../utils/errors.js';

const logger = createLogger('local-scan');

export interface LocalRepository {
  name: string;
  path: string;
  current_branch: string;
  default_branch?: string;
  remote_url: string | null;
  has_package_json: boolean;
  language: string | null;
}

export interface LocalReposResponse {
  repos: LocalRepository[];
  scan_path: string;
  total: number;
}

export class LocalScanService {
  /**
   * Scan a directory for git repositories (1 level deep)
   */
  async scanForRepos(basePath?: string): Promise<LocalReposResponse> {
    const scanPath = basePath || process.env.LOCAL_SCAN_DIR || this.getDefaultScanPath();

    logger.info('Scanning for local repos', { scanPath });

    const repos: LocalRepository[] = [];
    let totalDirsChecked = 0;

    try {
      const entries = await readdir(scanPath, { withFileTypes: true });
      totalDirsChecked = entries.length;

      for (const entry of entries) {
        if (!entry.isDirectory() || entry.name.startsWith('.')) continue;

        const dirPath = join(scanPath, entry.name);
        const gitPath = join(dirPath, '.git');

        try {
          await access(gitPath);
        } catch {
          continue; // No .git directory
        }

        try {
          const repo = await this.inspectRepo(dirPath, entry.name);
          repos.push(repo);
        } catch (error) {
          logger.warn('Failed to inspect repo', { dir: entry.name, error: getErrorMessage(error) });
        }
      }
    } catch (error) {
      logger.errorWithStack('Failed to scan directory', error as Error);
      throw error;
    }

    // Sort alphabetically
    repos.sort((a, b) => a.name.localeCompare(b.name));

    logger.info('Scan complete', { scanPath, reposFound: repos.length, totalDirsChecked });

    return { repos, scan_path: scanPath, total: repos.length };
  }

  /**
   * Inspect a single git repository
   */
  private async inspectRepo(repoPath: string, name: string): Promise<LocalRepository> {
    // Get current branch
    let currentBranch = 'main';
    try {
      currentBranch = execSync('git branch --show-current', { cwd: repoPath, encoding: 'utf-8' }).trim() || 'main';
    } catch {
      // fallback
    }

    // Detect default branch (prefer remote HEAD, fallback to current branch)
    let defaultBranch: string | undefined;
    try {
      const remoteHead = execSync('git symbolic-ref refs/remotes/origin/HEAD', { cwd: repoPath, encoding: 'utf-8' }).trim();
      // remoteHead looks like "refs/remotes/origin/main"
      if (remoteHead) {
        defaultBranch = remoteHead.replace('refs/remotes/origin/', '');
      }
    } catch {
      // No remote HEAD set - leave defaultBranch undefined, frontend will fallback to current_branch
    }

    // Get remote URL
    let remoteUrl: string | null = null;
    try {
      remoteUrl = execSync('git remote get-url origin', { cwd: repoPath, encoding: 'utf-8' }).trim() || null;
    } catch {
      // no remote
    }

    // Check for package.json
    let hasPackageJson = false;
    try {
      await access(join(repoPath, 'package.json'));
      hasPackageJson = true;
    } catch {
      // no package.json
    }

    // Detect language hint
    const language = await this.detectLanguage(repoPath, hasPackageJson);

    const result: LocalRepository = {
      name,
      path: repoPath,
      current_branch: currentBranch,
      remote_url: remoteUrl,
      has_package_json: hasPackageJson,
      language,
    };

    if (defaultBranch) {
      result.default_branch = defaultBranch;
    }

    return result;
  }

  /**
   * Simple language detection based on root files
   */
  private async detectLanguage(repoPath: string, hasPackageJson: boolean): Promise<string | null> {
    if (hasPackageJson) {
      // Check if it's TypeScript
      try {
        await access(join(repoPath, 'tsconfig.json'));
        return 'TypeScript';
      } catch {
        return 'JavaScript';
      }
    }

    // Check for other languages
    const languageFiles: [string, string][] = [
      ['Cargo.toml', 'Rust'],
      ['go.mod', 'Go'],
      ['requirements.txt', 'Python'],
      ['pyproject.toml', 'Python'],
      ['Gemfile', 'Ruby'],
      ['pom.xml', 'Java'],
      ['build.gradle', 'Java'],
      ['composer.json', 'PHP'],
    ];

    for (const [file, lang] of languageFiles) {
      try {
        await access(join(repoPath, file));
        return lang;
      } catch {
        continue;
      }
    }

    return null;
  }

  /**
   * Get the default scan path.
   * Finds the git workspace root from the user directory and goes one level up
   * so that sibling repos are visible.
   */
  private getDefaultScanPath(): string {
    const envDir = process.env['AGENT_BOARD_USER_DIR'];
    const baseDir = envDir || process.cwd();

    logger.info('Resolving default scan path', {
      baseDir,
      AGENT_BOARD_USER_DIR: envDir ?? '(not set)',
      source: envDir ? 'AGENT_BOARD_USER_DIR' : 'process.cwd()',
    });

    try {
      const gitRoot = execSync('git rev-parse --show-toplevel', {
        cwd: baseDir,
        encoding: 'utf-8',
      }).trim();

      logger.info('Git root detected', { gitRoot });

      // Go one level up from the git root so sibling repos are visible
      const scanPath = resolve(gitRoot, '..');

      logger.info('Default scan path resolved', { scanPath });

      return scanPath;
    } catch (error) {
      // Not inside a git repo — scan the directory itself
      logger.info('Git root detection failed, falling back to baseDir', {
        baseDir,
        error: getErrorMessage(error),
      });
      return baseDir;
    }
  }
}

// Singleton
let instance: LocalScanService | null = null;

export function getLocalScanService(): LocalScanService {
  if (!instance) {
    instance = new LocalScanService();
  }
  return instance;
}

File: packages/server-rs/src/services/local_scan_service.rs
//! Local scan service for discovering git repositories on the filesystem.
//!
//! Port of `packages/server/src/services/local-scan.service.ts`.
//!
//! Scans a base directory (one level deep) for subdirectories containing a `.git`
//! folder, then inspects each to extract basic repository metadata.

use std::path::{Path, PathBuf};
use std::process::Command as StdCommand;

use serde::{Deserialize, Serialize};
use tokio::fs;
use tracing::{debug, info, warn};

use crate::error::AppError;
use crate::services::stack_detector_service;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// A locally discovered git repository.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LocalRepository {
    pub name: String,
    pub path: String,
    pub current_branch: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_branch: Option<String>,
    pub remote_url: Option<String>,
    pub has_package_json: bool,
    pub language: Option<String>,
}

/// Response from scanning for local repositories.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LocalReposResponse {
    pub repos: Vec<LocalRepository>,
    pub scan_path: String,
    pub total: usize,
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Scans the given base path (one level deep) for git repositories.
///
/// If `base_path` is `None`, falls back to the `LOCAL_SCAN_DIR` environment
/// variable, then auto-detects from the git root of the current/user directory.
pub async fn scan_for_repos(base_path: Option<&Path>) -> Result<LocalReposResponse, AppError> {
    let scan_path = match base_path {
        Some(p) => p.to_path_buf(),
        None => get_default_scan_path(),
    };

    info!(scan_path = %scan_path.display(), "Scanning for local repos");

    let mut repos = Vec::new();

    let mut entries = fs::read_dir(&scan_path)
        .await
        .map_err(|e| AppError::Internal(anyhow::anyhow!("Failed to read directory: {e}")))?;

    while let Some(entry) = entries
        .next_entry()
        .await
        .map_err(|e| AppError::Internal(anyhow::anyhow!("Failed to read entry: {e}")))?
    {
        let file_type = match entry.file_type().await {
            Ok(ft) => ft,
            Err(_) => continue,
        };

        if !file_type.is_dir() {
            continue;
        }

        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if name_str.starts_with('.') {
            continue;
        }

        let dir_path = entry.path();
        let git_path = dir_path.join(".git");

        // Check for .git directory
        if fs::metadata(&git_path).await.is_err() {
            continue;
        }

        match inspect_repo(&dir_path, &name_str).await {
            Ok(repo) => repos.push(repo),
            Err(e) => {
                warn!(
                    dir = %name_str,
                    error = %e,
                    "Failed to inspect repo"
                );
            }
        }
    }

    // Sort alphabetically
    repos.sort_by(|a, b| a.name.cmp(&b.name));

    let total = repos.len();
    let scan_path_str = scan_path.to_string_lossy().to_string();

    info!(scan_path = %scan_path_str, repos_found = total, "Scan complete");

    Ok(LocalReposResponse {
        repos,
        scan_path: scan_path_str,
        total,
    })
}

// ---------------------------------------------------------------------------
// Internals
// ---------------------------------------------------------------------------

/// Inspects a single git repository directory to extract metadata.
async fn inspect_repo(repo_path: &Path, name: &str) -> Result<LocalRepository, AppError> {
    // Get current branch
    let current_branch = sync_git_output(&["branch", "--show-current"], repo_path)
        .unwrap_or_else(|| "main".to_string());

    // Detect default branch (prefer remote HEAD)
    let default_branch = sync_git_output(
        &["symbolic-ref", "refs/remotes/origin/HEAD"],
        repo_path,
    )
    .map(|head| head.replace("refs/remotes/origin/", ""));

    // Get remote URL
    let remote_url = sync_git_output(&["remote", "get-url", "origin"], repo_path);

    // Check for package.json
    let has_package_json = fs::metadata(repo_path.join("package.json")).await.is_ok();

    // Detect language
    let language = stack_detector_service::detect_language(repo_path, has_package_json).await;

    Ok(LocalRepository {
        name: name.to_string(),
        path: repo_path.to_string_lossy().to_string(),
        current_branch,
        default_branch,
        remote_url,
        has_package_json,
        language,
    })
}

/// Runs a synchronous git command and returns trimmed stdout, or `None` on failure.
///
/// We use `std::process::Command` (synchronous) for these one-shot queries because
/// they are fast and don't benefit from async I/O.
fn sync_git_output(args: &[&str], cwd: &Path) -> Option<String> {
    let mut cmd = StdCommand::new("git");
    cmd.args(args).current_dir(cwd);

    // On Windows, hide the console window
    #[cfg(target_os = "windows")]
    {
        use std::os::windows::process::CommandExt;
        const CREATE_NO_WINDOW: u32 = 0x08000000;
        cmd.creation_flags(CREATE_NO_WINDOW);
    }

    let output = cmd.output().ok()?;
    if !output.status.success() {
        return None;
    }
    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if stdout.is_empty() {
        None
    } else {
        Some(stdout)
    }
}

/// Determines the default scan path.
///
/// 1. Uses `LOCAL_SCAN_DIR` env var if set.
/// 2. Detects the git root of the current / user directory and goes one level up.
/// 3. Falls back to the current working directory.
fn get_default_scan_path() -> PathBuf {
    // Check env var
    if let Ok(dir) = std::env::var("LOCAL_SCAN_DIR") {
        if !dir.is_empty() {
            return PathBuf::from(dir);
        }
    }

    let base_dir = std::env::var("AGENT_BOARD_USER_DIR")
        .ok()
        .filter(|s| !s.is_empty())
        .map(PathBuf::from)
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));

    info!(base_dir = %base_dir.display(), "Resolving default scan path");

    // Try to get git root
    let mut cmd = StdCommand::new("git");
    cmd.args(["rev-parse", "--show-toplevel"]).current_dir(&base_dir);

    #[cfg(target_os = "windows")]
    {
        use std::os::windows::process::CommandExt;
        const CREATE_NO_WINDOW: u32 = 0x08000000;
        cmd.creation_flags(CREATE_NO_WINDOW);
    }

    if let Ok(output) = cmd.output() {
        if output.status.success() {
            let git_root = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if !git_root.is_empty() {
                let git_root_path = PathBuf::from(&git_root);
                // Go one level up so sibling repos are visible
                if let Some(parent) = git_root_path.parent() {
                    info!(scan_path = %parent.display(), "Default scan path resolved from git root");
                    return parent.to_path_buf();
                }
            }
        }
    }

    debug!(fallback = %base_dir.display(), "Git root detection failed, using base dir");
    base_dir
}
